### DISCUSSION TITLE
Does ruffle support "GPU path rendering"?

### DISCUSSION THREAD

**Discussion by Amaroq-Clearwater** [ruffle-rs/ruffle#16511] [General]
A topic of study that I've been following for a while (as in, years) was GPU-accelerated path rendering, potentially useful for things like SVGs (especially ones with animations and gradients, or even SVGs used as textures on 3D objects), Adobe Flash (or hey, Ruffle), and vector art programs like Adobe Illustrator.

Now that Ruffle is a thing, I figured it'd be a good idea to ask about it.

Here are just a few links on the subject--NOT an all-inclusive list:

- https://developer.nvidia.com/gpu-accelerated-path-rendering
- https://developer.nvidia.com/nv-path-rendering
- https://dl.acm.org/doi/10.1145/2980179.2982434

I'd love to hear some discussion on this!

**Comment by Fancy2209** [comment]
Yes! We render vectors with [Lyon](https://docs.rs/lyon/latest/lyon/)

**Comment by Lord-McSweeney** [comment]
Ruffle uses GPU rendering everywhere, except in the Canvas renderer on web. By default Ruffle will use the WebGL renderer on web, but if your device doesn't support WebGL, it will fall back to the Canvas renderer.

**Comment by torokati44** [comment]
> We render vectors with [Lyon](https://docs.rs/lyon/latest/lyon/)

So, no. 😶

> Ruffle uses GPU rendering everywhere, ...

True, but paths/curves always get tessellated on the CPU first.

**Reply by Amaroq-Clearwater** [reply]
That CPU tessellation thing sounds like a real buzzkill.

**Reply by torokati44** [reply]
Well it does indeed add a significant amount of startup delay on some content... And causes minor glitches on others. But it's hard to find a better method that is portable, can match FP's output, and isn't patent-encumbered...
There were some attempts though: https://github.com/ruffle-rs/ruffle/pull/814

**Reply by torokati44** [reply]
Not to mention: https://github.com/ruffle-rs/ruffle/pull/1168

**Reply by Amaroq-Clearwater** [reply]
> There were some attempts though: #814

Shame about Pathfinder…

**Comment by Dinnerbone** [comment]
I'd love one day for us to go back to vector graphics being a GPU thing, but sadly we're just not there :(

A few key points:
- Ruffle uses Flash's custom way of building paths, which doesn't at all match SVGs but with some preprocessing it can *almost* match it for most cases.
- To my knowledge, there's (not well supported) prototype opengl extensions to do *some kind of* Vector based graphics, but I don't think there's anything for vulkan/metal/dx12
- The platform that is hurt the most by the CPU based tessellation are the platforms guaranteed not to have these extensions: web and mobile

**Reply by Amaroq-Clearwater** [reply]
Even without GPU acceleration though... the rendering system could probably be improved a lot. Granted, I don't know how it works on a technical level, but I do know that there's always room for improvement.

For example, it would appear that many path renderers draw layers in order of back to front, instead of front to back, which means that if something is obscured... well, there's a reason 3D engines do front to back instead.

However, rendering front to back gets a little bit more complicated when transparency and layer blend modes enter the picture... and I don't have a solution to that particular issue.

**Reply by Fancy2209** [reply]
You are talking about vulkan/metal/dx12, but what about OpenVG?

**Reply by Amaroq-Clearwater** [reply]
Actually, I was talking about an existing extension of OpenGL
